File: Code for Java version of PRS/AlgoExec_M/FBKT.scala
package AlgoExec_M

import AlgoExec_W.FBKTWithinCore
import org.apache.spark.{HashPartitioner, SparkContext}
import SampleGen.SampleGenerate

import scala.collection.JavaConverters._
import scala.util.Random

object FBKT {
  /**
   * Function to select the best alternative using the FBKTPlus algorithm.
   * This function performs the FBKT selection process and returns the best alternative.
   *
   * @param ALT List of SampleGenerate objects representing the alternatives.
   * @param params Array of parameters required by the FBKTPlus algorithm.
   * @return The best alternative (SampleGenerate object) after selection.
   */
  private def select(ALT: List[SampleGenerate], params: Array[Double]): SampleGenerate = {
    val selection = new FBKTWithinCore(ALT.asJava, params)
    selection.FBKTSelect() // Perform FBKT selection
    selection.getBestAlt // Return the best alternative
  }

  /**
   * Function to generate a sample using the given parameters.
   * This function creates a new SampleGenerate object using the provided parameters.
   *
   * @param params Array of parameters required to generate the sample.
   * @return A new instance of SampleGenerate.
   */
  private def sampleGeneration(params: Array[Double]): SampleGenerate = {
    new SampleGenerate(params) // Create a new SampleGenerate instance
  }

  /**
   * Function to generate samples from input parameters, with or without using a random seed.
   *
   * This function processes an RDD of strings, where each string contains an ID and a set of parameters.
   * It then generates samples using either a seeded random number generator or without seeding,
   * depending on the useSeed flag. The generated samples are paired with their corresponding IDs.
   *
   * @param sampleParam An RDD of strings, where each string represents an ID and a set of parameters, space-separated.
   * @param seed An integer seed used to initialize the random number generator for reproducibility.
   * @return An RDD of tuples, where each tuple contains an ID (Int) and a SampleGenerate object.
   */
  private def generateSample(sampleParam: org.apache.spark.rdd.RDD[String], seed: Int): org.apache.spark.rdd.RDD[(Int, SampleGenerate)] = {
    // Map over each line in the RDD and generate samples with a seeded random number generator
    sampleParam.map { line =>
      val (id, param) = parseLine(line) // Parse the line into an ID and parameters
      val random = new Random(seed + id) // Initialize a new random generator with a seed unique to the ID
      val altSeed = random.nextInt() // Generate an alternative seed using the random generator
      val alt = sampleGeneration(param) // Generate a sample using the parameters
      alt.setSeed(altSeed) // Set the seed for the generated sample to ensure reproducibility
      (id, alt) // Return a tuple containing the ID and the generated sample
    }
  }

  /**
   * Function to parse a line of input text into an ID and an array of parameters.
   *
   * This function takes a line of input, splits it into components, converts the first element
   * into an integer ID and the remaining elements into a double array of parameters.
   *
   * @param line A line of input text in the format "ID param1 param2 ... paramN".
   * @return A tuple containing:
   *         - ID: An integer representing the ID.
   *         - params: An array of doubles representing the parameters.
   */
  private def parseLine(line: String): (Int, Array[Double]) = {
    val param = line.split(" ") // Split the line by spaces to separate the ID and parameters
    val id = param(0).toInt // Convert the first element to an integer ID
    val params = param.map(_.toDouble) // Convert the entire array to doubles (ID will also be converted, but only the parameters are used)
    (id, params) // Return a tuple containing the ID and the parameters array
  }

  /**
   * Function to run the FBKT algorithm.
   * This function orchestrates the FBKT process by reading input data, generating samples,
   * partitioning them, and finally selecting the best alternative.
   *
   * @param sc SparkContext used for distributed computing.
   * @param paramOfSelect Array of parameters required by the FBKT process.
   * @param nCores Number of CPU cores to use.
   * @param inputPath Path to the input data file.
   * @param seed Seed for random number generation.
   * @param repeat Number of repetitions of the FBKT process.
   * @return A sequence of tuples containing:
   *         - Int: Replication number.
   *         - Int: ID of the best alternative selected.
   *         - Double: Wall-clock time in seconds.
   *         - Double: Total simulation time in seconds.
   */
  def runFBKT(sc: SparkContext, paramOfSelect: Array[Double], nCores: Int, inputPath: String, seed: Int, repeat: Int): Seq[(Int, Int, Double, Double)] = {
    val results = new scala.collection.mutable.ArrayBuffer[(Int, Int, Double, Double)]() // Buffer to store results
    var count = 0 // Initialize repetition counter
    val randomSeedGenerator = new Random(seed) // Initialize random seed

    // Repeat the FBKT algorithm for the specified number of times
    while (count < repeat) {
      val startTime = System.nanoTime() // Record start time
      val repeatSeed = randomSeedGenerator.nextInt() // Generate a common random number

      // Read input data and generate samples
      val sampleParam = sc.textFile(inputPath, nCores)
      val sample = generateSample(sampleParam, repeatSeed)

      // Partition the samples and cache them
      val uniformSampleKey = sample.partitionBy(new HashPartitioner(nCores))
      val uniformSample = uniformSampleKey.map(_._2).glom().map(_.toList)

      // Perform the FBKT selection
      val bestAltCore = uniformSample.map(ALT => select(ALT, paramOfSelect))

      // Reduce to find the best alternative
      val bestAlt = bestAltCore.reduce { (x, y) =>
        if (x.getMean > y.getMean) {
          x.setN(x.getN + y.getN) // Update sample size
          x.setSimulateTime(x.getSimulateTime + y.getSimulateTime) // Update simulation time
          x // Return x if its mean is higher
        } else {
          y.setN(y.getN + x.getN) // Update sample size
          y.setSimulateTime(y.getSimulateTime + x.getSimulateTime) // Update simulation time
          y // Return y if its mean is higher
        }
      }

      // Collect results
      val replication = count + 1 // Increment replication number
      val bestID = bestAlt.getId // Get ID of the best alternative
      val wallClockTime = (System.nanoTime() - startTime) / 1e9 // Calculate wall clock time in seconds
      val totalSimulateTime = bestAlt.getSimulateTime // Get total simulation time in seconds

      results += ((replication, bestID, wallClockTime, totalSimulateTime)) // Add results to buffer
      count += 1 // Increment counter
    }
    // Return the results as a sequence
    results
  }
}

File: Code for Python version of PRS/AlgoExec_M/FBKT.py
import time
import numpy as np
from pyspark import SparkConf, SparkContext
from AlgoExec_W.FBKTWithinCore import FBKTPlus
from SampleGen.SampleGenerate import SampleGenerate


def select(ALT, Params):
    """
    Perform selection of the best alternative using the FBKTPlus algorithm.

    Args:
        ALT (list): List of alternative systems to be evaluated.
        Params (list): Parameters required for the FBKTPlus selection procedure.

    Returns:
        SampleGenerate: The best alternative system selected by the FBKTPlus algorithm.
    """
    selection = FBKTPlus(ALT, Params)  # Initialize FBKTPlus with alternatives and parameters
    selection.FBKTselect()  # Execute the FBKT selection procedure
    return selection.get_best_alt()  # Return the best alternative identified


def sample_generation(params):
    """
    Generate a sample based on the provided parameters.

    Args:
        params (np.array): Parameters required to generate the sample.

    Returns:
        SampleGenerate: An instance of SampleGenerator containing the generated alternative.
    """
    alt = SampleGenerate(params)  # Create a new SampleGenerator instance using the given parameters
    return alt


def reduce_logic(x, y):
    """
    Combine two SampleGenerator objects based on their mean values and simulation data.

    Args:
        x (SampleGenerate): The first SampleGenerator object to compare.
        y (SampleGenerate): The second SampleGenerator object to compare.

    Returns:
        SampleGenerate: The SampleGenerator object with the higher mean value,
                         updated with the combined sample size and simulation time.
    """
    if x.get_mean() > y.get_mean():  # Compare mean values of x and y
        x.set_N(x.get_N() + y.get_N())  # Add the sample sizes of x and y
        x.set_simulation_time(x.get_simulation_time() + y.get_simulation_time())  # Add the simulation times
        return x  # Return x if it has the higher mean
    else:
        y.set_N(x.get_N() + y.get_N())  # Add the sample sizes of y and x
        y.set_simulation_time(y.get_simulation_time() + x.get_simulation_time())  # Add the simulation times
        return y  # Return y if it has the higher mean


def generate_sample(sample_param, random_seed):
    """
    Generates samples based on input parameters, with or without a random seed.

    Args:
        sample_param: An iterable (e.g., RDD or list) containing lines of data, where each line is a space-separated
                      string with an ID followed by parameters.
        random_seed: An integer seed used for random number generation to ensure reproducibility.

    Returns:
        An iterable (e.g., RDD or list) of tuples, where each tuple contains an ID and a generated alternative.
    """

    def parse_line(line):
        """
        Parses a line of input data to extract the ID and parameters.

        Args:
            line: A string where the first element is the ID and the rest are parameters, space-separated.

        Returns:
            A tuple containing:
                - ID: An integer representing the ID.
                - params: A NumPy array of floats representing the parameters.
        """
        parts = line.split(" ")
        ID = int(parts[0])  # Extract ID as an integer
        params = np.array(list(map(float, parts)))  # Convert the rest of the line to a NumPy array of floats
        return ID, params

    def generate_sample_with_seed(line, in_seed):
        """
        Generates a sample using a random seed for reproducibility.

        Args:
            line: A string of input data to parse and generate a sample from.
            in_seed: An integer seed to initialize the random number generator.

        Returns:
            A tuple containing:
                - ID: An integer representing the ID.
                - alt: The generated alternative using seeded random number generation.
        """
        ID, params = parse_line(line)
        random = np.random.RandomState(in_seed + ID)  # Initialize a random generator with a unique seed
        alt_seed = random.randint(0, np.iinfo(np.int32).max)  # Generate an alternative seed within the int32 range
        alt = sample_generation(params)  # Generate the alternative
        alt.set_seed(alt_seed)  # Set the alternative seed
        return ID, alt  # Return the ID and generated alternative

    # Use the appropriate sample generation function based on whether a seed should be used
    return sample_param.map(lambda line: generate_sample_with_seed(line, random_seed))


def run_fbkt_process(N, n0, phi, nCores, parameters_path, seed=1234, repeat=1):
    """
    Main function to run the FBKTPlus algorithm with the specified parameters.

    Args:
        N (int): Total sample budget to be allocated across cores.
        n0 (int): Initial number of samples for each alternative.
        phi (int): Parameter that determines budget allocation in the FBKTPlus algorithm.
        nCores (int): Number of CPU cores to use for parallel processing.
        parameters_path (str): Path to the input data file containing alternative parameters.
        seed (int, optional): Random seed for reproducibility. Defaults to -1 (no fixed seed).
        repeat (int, optional): Number of times to repeat the FBKT process. Defaults to 1.

    Returns:
        list: A list of tuples containing the results of each replication, where each tuple includes:
              - Replication number (int)
              - ID of the best alternative (int)
              - Wall clock time (float)
              - Total simulation time for the best alternative (float)
    """
    # Initialize Spark configuration and create or get the SparkContext
    conf = SparkConf().setAppName("FBKT").set("spark.cores.max", str(nCores))
    sc = SparkContext.getOrCreate(conf)  # Get or create a SparkContext

    # Define parameters for the FBKTPlus algorithm
    params = [int(N / nCores), n0, phi]

    results = []  # List to store the results of each replication

    # Initialize random seed generator
    random_repeat_seed = np.random.RandomState(seed)

    def run_fbkt(repeat_seed, replication):
        """
        Inner function to execute the FBKTPlus algorithm for a single replication.

        Args:
            repeat_seed (int): Seed for random number generation.
            replication (int): Replication number for tracking purposes.
        """
        # Load input data and generate samples
        sample_param = sc.textFile(parameters_path, nCores)
        sample = generate_sample(sample_param, repeat_seed)

        # Partition the samples and group them for processing
        uniform_sample_key = sample.partitionBy(nCores)
        uniform_sample = uniform_sample_key.map(lambda x: x[1]).glom().map(lambda x: list(x))

        # Perform FBKT selection on each partition and reduce to find the best alternative
        best_alt_core = uniform_sample.map(lambda ALT: select(ALT, params))
        best_alt = best_alt_core.reduce(lambda x, y: reduce_logic(x, y))

        # Get details of the best alternative
        best_args = best_alt.get_args()
        clock_time = time.time() - start_time  # Calculate wall clock time
        total_simulation_time = best_alt.get_simulation_time()  # Get the total simulation time

        # Store the results of the current replication
        results.append((replication + 1, int(best_args[0]), clock_time, total_simulation_time))

    # Run the FBKTPlus algorithm for the specified number of repetitions
    for _ in range(repeat):
        start_time = time.time()  # Record start time
        run_fbkt(random_repeat_seed.randint(0, np.iinfo(np.int32).max), _)  # Run FBKT for the current repeat

    sc.stop()  # Stop the SparkContext after all repetitions are complete
    return results  # Return the list of results
